lib/src/components/NxFileUpload/__tests__/NxFileUpload.test.tsx

```diff
@@ -0,0 +1,80 @@
+      renderInput = pipe(renderEl, el => el?.querySelector('input[type=file]'));
+
+  it('renders a file input and a button', function() {
+    const component = render(),
+
+        // file inputs have no aria role and the label text is outside of the scope of this component,
+        // so we have to query like this
+        input = component.container.querySelector('input[type=file]'),
+        button = component.queryByRole('button', { name: 'Choose File' });
+
+    expect(input).toBeDefined();
+    expect(button).toBeDefined();
+
+    // cannot effectively test that the button activates the input here, will do in functional test
+  });
+
+  it('renders a "No file selected" message if files is null', function() {
+    expect(renderEl()).toHaveTextContent('No file selected');
+  });
+
+  it('shows an error when there is no file selected, if it isRequired is not isPristine', function() {
+    expect(render().queryByRole('alert')).not.toBeTruthy();
+    expect(render({ isPristine: true }).queryByRole('alert')).not.toBeTruthy();
+    expect(render({ isRequired: true, isPristine: true }).queryByRole('alert')).not.toBeTruthy();
+
+    expect(render({ isRequired: true, isPristine: false }).queryByRole('alert'))
+        .toHaveTextContent('This field is Required!');
+  });
+
+  it('sets aria-required on the input if isRequired is true', function() {
+    expect(renderInput()?.getAttribute('aria-required')).not.toBe('true');
+    expect(renderInput({ isRequired: false })?.getAttribute('aria-required')).not.toBe('true');
+    expect(renderInput({ isRequired: true })?.getAttribute('aria-required')).toBe('true');
+  });
+
+  it('sets aria-invalid on the input when there is no file selected, if it isRequired is not isPristine', function() {
+    expect(renderInput()?.getAttribute('aria-invalid')).not.toBe('true');
+    expect(renderInput({ isPristine: true })?.getAttribute('aria-invalid')).not.toBe('true');
+    expect(renderInput({ isRequired: true, isPristine: true })?.getAttribute('aria-invalid')).not.toBe('true');
+
+    expect(renderInput({ isRequired: true, isPristine: false })?.getAttribute('aria-invalid')).toBe('true');
+  });
+
+  it('sets aria-errormessage on the input to the id of the validation error, when present', function() {
+    expect(renderInput()).not.toHaveErrorMessage();
+    expect(renderInput({ isPristine: true })).not.toHaveErrorMessage();
+    expect(renderInput({ isRequired: true, isPristine: true })).not.toHaveErrorMessage();
+
+    expect(renderInput({ isRequired: true, isPristine: false })).toHaveErrorMessage('This field is Required!');
+  });
+
+  it('attaches a ref to the top-level element', function() {
+    const ref = React.createRef<HTMLDivElement>(),
+        renderedEl = renderEl({ ref });
+
+    expect(ref.current).toBe(renderedEl);
+  });
+
+  // Everything that involves actually selecting a file cannot be cleanly tested in RTL, will be done in
+  // functional tests
```

Comment:
Out of curious, would you please elaborate more why it can't be clean if tested in RTL? Thanks.

Reply:
> Out of curious, would you please elaborate more why it can't be clean if tested in RTL? Thanks.

The only thing that RTL has for interaction with file upload components is https://testing-library.com/docs/user-event/utility#upload .  That utility is inadequate for our needs for two reasons:
1. It simulates a user clicking on the `<input>` itself.  In `NxFileUpload` the `<input>` is invisible and a real user actually triggers it by clicking the adjacent `<button>`.  So simulating a click on the `<input>` doesn't really verify anything with real-world relevance.
2. The `FileList` object that it results in is fake.  It's a plain JS object with all the the properties that you'd expect from a `FileList` and with its prototype hacked using `Object.setPrototypeOf`.  This is adequate for most cases of application code interacting with the `FileList` but it is _not_ adequate for any case where that object gets reassigned back to the `<input>`'s `value` property – which is exactly what `NxFileUpload` does.  In fact, `<input>`s are necessarily extremely strict about ensuring that it is impossible for a JS-constructed "FileList" to ever be accepted by the `value` property.  This is a security feature: the only way to get a reference to a file on the user's system is by the real user consciously selecting file via their interaction with the file input.  It must remain impossible for some JavaScript code to trigger an upload of, say, `/etc/passwd` without the user's direction.

Reply:
Thanks for the clarification.